### PR TITLE
KT-6217 Add inspection that reports unused equals expression

### PR DESCRIPTION
**Pull request by shiraji**
#KT-6217 Fixed

https://youtrack.jetbrains.com/issue/KT-6217

### CODE REVIEW

File: idea/resources/inspectionDescriptions/UnusedEquals.html [hunk 1]
```diff
@@ -0,0 +1,5 @@
+<html>
+<body>
+This inspection reports unused equals expression.
```
Comment (cypressious) [inline idea/resources/inspectionDescriptions/UnusedEquals.html]: expression**s**

File: idea/src/org/jetbrains/kotlin/idea/inspections/UnusedEqualsInspection.kt [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor {
+        return object : KtVisitorVoid() {
+            override fun visitBinaryExpression(expression: KtBinaryExpression) {
+                super.visitBinaryExpression(expression)
+                if (expression.operationToken != KtTokens.EQEQ) return
+                val parent = expression.parent as? KtBlockExpression ?: return
+                val lastBlockStatement = parent.lastBlockStatementOrThis()
+                when {
+                    expression.evaluatesTo(lastBlockStatement) && expression.getLineNumber() == lastBlockStatement.getLineNumber() ->
+                        expression.getStrictParentOfType<KtLambdaExpression>()?.visitLambdaExpression(holder, expression) ?: holder.registerUnusedEqualsProblem(expression)
```
Comment (mglukhikh) [inline idea/src/org/jetbrains/kotlin/idea/inspections/UnusedEqualsInspection.kt]: All of this looks too nasty and unsafe. Please use `KtExpression.isUsedAsExpression` from `BindingContextUtils.kt` to determine whether your `equals` result is used or not.

File: idea/src/org/jetbrains/kotlin/idea/inspections/UnusedEqualsInspection.kt [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+                        if (!expression.evaluatesTo(lastBlockStatementOrThis)) holder.registerUnusedEqualsProblem(expression)
+                    }
+                    else -> holder.registerUnusedEqualsProblem(expression)
+                }
+            }
+
+            private fun ProblemsHolder.registerUnusedEqualsProblem(expression: KtBinaryExpression) {
+                registerProblem(expression,
+                                "Unused equals expression",
+                                ProblemHighlightType.LIKE_UNUSED_SYMBOL)
```
Comment (mglukhikh) [inline idea/src/org/jetbrains/kotlin/idea/inspections/UnusedEqualsInspection.kt]: Should we provide some quick fix? I'd say that in simple cases with local variables or constants binary expression could be completely deleted, but in cases like `foo() == bar()` we should get

```
foo()
bar()
```

instead

File: idea/src/org/jetbrains/kotlin/idea/inspections/UnusedEqualsInspection.kt [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+                    KotlinBuiltIns.isBoolean(lambdaTypeArguments[0].type) -> {
+                        val lastBlockStatementOrThis = bodyExpression?.lastBlockStatementOrThis() ?: return
+                        if (!expression.evaluatesTo(lastBlockStatementOrThis)) holder.registerUnusedEqualsProblem(expression)
+                    }
+                    else -> holder.registerUnusedEqualsProblem(expression)
+                }
+            }
+
+            private fun ProblemsHolder.registerUnusedEqualsProblem(expression: KtBinaryExpression) {
+                registerProblem(expression,
```
Comment (mglukhikh) [inline idea/src/org/jetbrains/kotlin/idea/inspections/UnusedEqualsInspection.kt]: I'd say it's better to provide also text range here and highlight `==` only as unused.

**Comment by elizarov** [comment]
I'd suggest that in cases like "a == something" where a is some modifiable thing the most likely explanation is that it was a typo and a fix is to remove an extra =

**Comment by cypressious** [comment]
Maybe all the quick fixes are a bit to speculative and not really
necessary. Having the warning is the most important thing.

Kirill Rakhman

Am 23.12.2016 12:27 nachm. schrieb "Roman Elizarov" <
notifications@github.com>:

**Comment by elizarov** [comment]
Good point.

**Comment by mglukhikh** [comment]
Ok, quick-fixes are really controversial here so we can omit them.

**Comment by shiraji** [comment]
Done. Thanks for pointing them. 

At the first I was thinking a quick fix and notice there are so many ways to fix the expressions. As @cypressious https://github.com/JetBrains/kotlin/pull/1000#issuecomment-268977235, I decided not to provide them.

**Comment by mglukhikh** [comment]
`isUsedAsResultOfLambda` is not the only case. What about something like `val eq = if (...) a == b else c == d`, or the same with braces: `val eq = if (...) { a == b } else { c == d }`?

**Comment by shiraji** [comment]
Done for the case that the binary expression is used as an expression.

**Comment by mglukhikh** [comment]
Merged manually into master

**Pull request closed without merging**